build: do not drop commits with malformed messages in the release note

wookayin opened the pull request:
Problem: Some commits that do not follow the conventional commits format
are not included in the release noted generated by `git cliff`.
An example: `[Backport release-0.9] fix(build): ...`

While CI enforces "lintcommit" checking, such malformed commits might
have been merged via merging "backport to a release branch" through
github, where (accidentally) the commit message gets rewritten and no
longer follows the conventional commits format.

Solution:

- Strip the `[Backport release-..]` prefix while pre-processing commits,
  because this is a common error case in our current release process
  (applies to the commits that were already merged).

- Set `filter_commits = false` and `filter_unconventional = false`, i.e.
  do not filter out the commits whose message do correctly follow the
  conventional commits. This will make mistakes and errors at least
  *recognizable* while generating a release note during release
  (which should, however, be manually fixed by the release manager).
  NOTE: This changes will also result in new "vim-patch" section.

- Include "ci" groups which were previously missing.

wookayin commented:
Changes for release note v0.9.4..v0.9.5:

```
--- CHANGELOG.md	2023-12-31 05:17:55
+++ changelog2.md	2023-12-31 05:18:14
@@ -8,6 +8,7 @@
 - Dont create data dir if it's a broken symlink
 - Make InspectTree handle nested injection
 - Remove nested for_each_tree in TSTreeView (#26331)
+- Vim.treesitter.get_node() now correctly takes opts.lang (#26382)
 - **api**: Use a conditional stack for nvim_cmd
 - **api**: Don't set coladd of mark
 - **change**: Update fold after on_bytes
@@ -17,6 +18,7 @@
 - **highlight**: Apply 'winblend' to float border
 - **inccommand**: Save and restore '[ and '] marks
 - **inccommand**: Don't crash with "split" and 'n' flag
+- **log**: Increase size of buffer for nvim instance name (#26450)
 - **lsp**: Handle NUL bytes in popup text
 - **lua**: Correct return value for on_key with no arguments
 - **lua**: Crash in nlua_error
@@ -62,6 +64,10 @@
 - **starting.txt**: Correct step number
 
 
+### Features
+- **highlight**: Allow hyphens (-) in highlight group names (#25661)
+
+
 ### Performance
 - Remove redundant strlen in skipwhite
 
@@ -77,4 +83,17 @@
 - **terminal/channel_spec**: Fix flakiness
 
 
+### Vim-patch
+- 9.0.0064: confusing error when using "q:" in command line window
+- 9.0.0218: reading before the start of the line
+- 9.0.0249: no test for what 9.0.0234 fixes
+- 9.0.0490: using freed memory with cmdwin and BufEnter autocmd
+- 9.0.0492: cmdwin test fails on MS-Windows
+- 9.0.0598: using negative array index with negative width window
+- 9.0.2151: 'breakindent' is not drawn after diff filler lines
+- Ff0baca86523
+- 9.0.2159: screenpos() may crash with neg. column
+- 9.0.2178: reg_executing() wrong for :normal with range
+
+
 <!-- generated by git-cliff -->
```

bfredl commented:
It is a simple workaround _after the fact_ for `release-0.9` but I'm not sure I like the precedent it sets (and thus, adding also to the _master_ branch like this PR does). The risk is that we would just pile on more exceptions, instead of making sure to actually follow the intentional format in the first place.

So, specifically about backports.  If there is a general expectation that PR titles of _autogenerated_ PR:s (such as backport PR:s) are good to use, then I think we should not prepend `[Backport release-0.9]` to such PR titles in the first place, and instead make them standout in some other way (like screaming label colors). but such PR:s get merged very quickly, so there might be no problem with cluttering the open PR view in the first place. Even if PR titles are not validated in general I understand the expectation that auto-generated messages are good, because why would we have generated a bad message and not a good message

wookayin commented:
> ... instead of making sure to actually follow the intentional format in the first place.

Yes! We should make all the commits merged to HEAD (`master`) and backport release branches follow the correct commit message format -- as much as possible. However, unless everything is automated and *strictly* enforced, there still might be some exceptions or human mistakes happen. Devs often make use of Github's feature to "edit the commit message" while merging PRs through web UI (to correct typos or fix commit messages at ease), I'm a bit against too-strict enforcement. Nevertheless, we can consider further improvements to the PR backport automation (e.g. do not have the "[backport release]" prefix, but making it as a suffix, etc.).

A minor remark:

Exceptions --- #26800 (by @zeertzjq) also draws my attention because it cherry-picks two commits from deps/libtermkey that violate the intended commit message format. It's debatable whether we want these two commits ( b4ef9134007aeab421d93299bdf3b97fba0ffa79, ffe96c6c1c6c54bc03f2d534c9a6bf92c2dd1e51) include in the release note, and whether these would be manually edited for the release note during the release process. However, I think being aware of some malformed commit messages would be still beneficial because otherwise `git cliff` will not generate any warnings or notifications.

wookayin committed:
build: do not drop commits with malformed messages in the release note

Problem: Some commits that do not follow the conventional commits format
are not included in the release noted generated by `git cliff`.
An example: `[Backport release-0.9] fix(build): ...`

While CI enforces "lintcommit" checking, such malformed commits might
have been merged via merging "backport to a release branch" through
github, where (accidentally) the commit message gets rewritten and no
longer follows the conventional commits format.

Solution:

- Strip the `[Backport release-..]` prefix while pre-processing commits,
  because this is a common error case in our current release process.
  (this applies to the commits that have already been merged)

- Set `filter_commits = false` and `filter_unconventional = false`, i.e.
  do not filter out the commits whose message do correctly follow the
  conventional commits. Such commits will be categorized as a separate
  group at the end. This will make mistakes and errors at least
  *recognizable* while generating a release note during release
  (which should be manually but easily fixed by the release manager).
  NOTE: This changes will also result in new "vim-patch" section.

- Also, include "ci" groups which were previously missing.

bfredl commented:
> I'm a bit against too-strict enforcement. 

I am as well, but this means we need to accept that generated and edited release notes are going to contain some  errors and omissions as well. Getting rid of _systematic_ sources of errors (i e unfortunate choices in automated scripts) is going to eliminate a lot of unnecessary omissions. Then we can reassess the situation after the next release:)

We can make the process of release notes more open, like some other projects. I e make the notes available as a draft for others to peek at and suggest changes. Although this is going to make the release process more involved. It might be something to consider for feature releases at some point (but not for patch releases, the less extra steps the better).

justinmk commented:
> We can make the process of release notes more open, like some other projects. I e make the notes available as a draft for others to peek at and suggest changes.

Based on past newsletter PRs, I don't expect much non-core feedback on such a PR.

> The risk is that we would just pile on more exceptions, instead of making sure to actually follow the intentional format in the first place.

100%. The purpose of cliff is to help produce release notes. And the purpose of conventional commits is to help automate that. `filter_unconventional = false` seems to defeat much of what was gained.

For maintenance releases, can't we just link to the commit history? Changelog isn't as important there.

wookayin commented:
Thanks for your comment! Let me point out again that the whole point of this is to prevent commits or release note items from being excluded from the release note. Automation and less manual jobs are of course good, but I think we should aim for correctness too.

> The purpose of cliff + conventional commits is to reduce the work involved in producing release notes.

If we want *full* automation while ensuring correctness (i.e., not unawarely missing any commits), we probably would need have better enforcement. Because of many possible reasons and the current limitation of Github UI/UX, conventional commits are not strictly enforced; (1) a maintainer can change title (on squashing merge) even if lintcommit has passed and make some mistakes, (2) commits in PR passed CI but the "title of a PR" had a mistake so the merged commit ended up violating conventional-commit (e.g. #26807), to name a few. Indeed, for upcoming v0.10 release (between v0.9.0..HEAD), there are already a few commits that violate the commit message format. However, we probably don't want to make the process of merging PR unnecessarily complicated.

If we want to get rid of any manual process in writing the git-cliff-generated release note, we could just have a *"uncategorized"* or "others" section collecting all the unclassified (incorrectly-formatted) commits instead of manually editing the release note (which shouldn't take too much time, to delete or move around only a few lines).

justinmk commented:
> we could just have a _"uncategorized"_ or "others" section collecting all the unclassified (incorrectly-formatted) commits

I guess that's ok. That's this part, right?:

    { message = "^.*", group = "(unclassified; TODO: review manually)"},

justinmk commented on scripts/cliff.toml:
```diff
@@ -56,11 +62,16 @@ commit_parsers = [
-# filter out the commits that are not matched by commit parsers
-filter_commits = true
+# do not filter out the commits that are not matched by commit parsers
+filter_commits = false
```
let's move this next to `filter_unconventional`

justinmk replied:
is this change actually needed? wouldn't the `message = "^.*"` commit parser match everything?

wookayin commented:
> So that means the release manager can easily delete that whole section, and then it's equivalent to before?

Yes, exactly. I've made a few additional changes so now the "uncategorized" section would come at the end. The new sections are vim patches (not sure we want to include this or not because it's quite messy and looooong) and this uncategorized section; if one removes these two these should be equivalent (except for other changes regarding "[backport release]").

At the moment, I can see around 5 minor commits as well as `version.c: Update` commits (`git cliff --config scripts/cliff.toml v0.9.0..HEAD`). BTW there is [another issue](https://github.com/orhun/git-cliff/issues/423) as a side effect but hope this can be addressed by upstream (git-cliff) fix.

justinmk commented on scripts/cliff.toml:
```diff
@@ -38,16 +38,27 @@ footer = """
 commit_parsers = [
+    # Exclude "merge commits"
+    { message = '^Merge pull request #[0-9]+ from .*[\n\s]*', skip = true },
+    { message = '^Merge #[0-9]+', skip = true },
```
would it be more reliable to filter merge commits like this: https://git-cliff.org/docs/tips-and-tricks#filter-merge-commits

i.e. change line 17 to : 

    {% for group, commits in commits | group_by(attribute="group") | filter(attribute="merge_commit", value=false) %}

justinmk commented on scripts/cliff.toml:
```diff
@@ -38,16 +38,27 @@ footer = """
 ]
 # regex for parsing and grouping commits
 commit_parsers = [
+    # Exclude "merge commits"
+    { message = '^Merge pull request #[0-9]+ from .*[\n\s]*', skip = true },
+    { message = '^Merge #[0-9]+', skip = true },
```
would it be more reliable to filter merge commits like this: https://git-cliff.org/docs/tips-and-tricks#filter-merge-commits

i.e. change line 17 to : 

    {% for group, commits in commits | group_by(attribute="group") | filter(attribute="merge_commit", value=false) %}

wookayin replied:
Great suggestion. Will do that

justinmk commented on scripts/cliff.toml:
```diff
@@ -56,11 +67,16 @@ commit_parsers = [
+
+    # All other commits that are not conventional should be reviewed manually
+    # Note: '[' is lexicographically larger than 'Z', so this section is put at the end
+    { message = "^([^)\r\n]+):", group = "[unclassified; TODO: review manually]", scope = "$1" },
```
```suggestion
    { message = "^([^)\r\n]+):", group = "unclassified; review manually, update cliff.toml if necessary", scope = "$1" },
```

dundargoc commented:
[This PR](https://github.com/neovim/neovim/pull/28590) will remove the `[Backport release-0.9]` from being created in pull request titles, so we don't need to explicitly exclude it anymore.

dundargoc commented:
Superseded by https://github.com/neovim/neovim/pull/28595.

I changed the strategy from including non-conventional commits to just yeeting them as I don't think we have any automated solutions left that create non-conventional commits.
